### PR TITLE
x86: support single or per-domain page tables

### LINKED ISSUE #29175: x86 fails all tests if CONFIG_X86_KPTI is disabled
**Describe the bug**
If kernel page table isolation (KPTI) is disabled, the memory management code doesn't work. This may have been bit-rotten for a while, we were lacking test coverage.

**To Reproduce**
This PR illustrates the problem:
https://github.com/zephyrproject-rtos/zephyr/pull/28800

Build tests with the new `qemu_x86_nokpti` or `qemu_x86_64_nokpti` targets.

**Expected behavior**
All tests pass.

**Impact**
x86 CPUs not susceptible to Meltdown cannot disable this feature and enjoy the considerable performance benefits.

### LINKED ISSUE #29382: remove memory domain restriction on system RAM for memory partitions on MMU devices
**Is your enhancement proposal related to a problem? Please describe.**
MMU devices have a documented restriction that k_mem_partitions must be within system RAM bounds in order to be programmed into a memory domain.

This is overly restrictive. We should allow k_mem_partition to be installed for any mapped memory address, including virtual mappings from z_mem_map().

The immediate use-case is allowing for user-mode drivers that can be granted access to MMIO regions.

**Describe the solution you'd like**
Documentation removed that memory domain partitions can only be SRAM address ranges.
x86 implementation updated to support this.
TBD: can we provide a generic test?

### PR DESCRIPTION
**Pull request by tyrantlizard77**
This PR contains many updates towards getting the x86 implementation to match our plans in https://github.com/zephyrproject-rtos/zephyr/issues/27819. Major updates:

The x86 port used to maintain a separate set of page tables for every user thread. This was ok for 64-bit use-cases where RAM was not much of a concern, but this isn't workable for microcontroller use-cases. This PR implements one of the two page table management schemes described in #27819 
 - By default, page tables are now maintained at the memory domain level, with threads in the same memory domain sharing the same page tables.
 - If `CONFIG_X86_COMMON_PAGE_TABLE` is enabled, there is instead a single set of page tables for all threads/domains that is updated on context switch.
 - For either page table management scheme, x86 allows memory domains to program partitions for any mapped address, not just permanent RAM mappings
 - X86 now implements `struct arch_mem_domain` with page table information if multiple page tables are in use
 - X86 now implements `arch_mem_domain_init()` which creates a fresh set of page tables for the new memory domain, copied from the boot page tables and using the page pool for memory, if we are not using the common page table
 - If multiple page tables are in use, memory mappings are the same across all page tables and all threads. Only permissions with respect to user threads and their memory domain membership is different. Memory access to supervisor mode is the same regardless what thread is scheduled.
 - Since the same page tables could be active on multiple CPUs, implement TLB shootdown logic for SMP.
 - The `z_x86_userspace_enter` implementation now calls back into C code in order to wipe the target stack and update the page table permissions, after switching onto the privilege elevation stack. The US bit for the stack region can only be activated after it has been cleared for security reasons, but before the thread drops privileges.
 - A spinlock now protects the set of active page tables and the slist that links them. All page table modifications are serialized on this lock.
 - We use unused bits in the PTE to backup original page permissions when the mapping was made. This lets us easily restore them when un-applying memory domain partitions.
 - Added build targets to exercise configurations where KPTI is disabled, thanks to @laurenmurphyx64 for the patches
 - Added a test scenario specifically to test memory domain migration, important for SMP if the target thread is running on another CPU.
 - Added qemu_x86_tiny build target, which exercises `CONFIG_X86_COMMON_PAGE_TABLE` with the page pool disabled and a 4MB address space

Fixes: #29175 
Fixes: #29382
Related to: #27819

### CODE REVIEW

**Comment by andrewboie** [comment]
Updated PR. Includes test case changes from https://github.com/zephyrproject-rtos/zephyr/pull/29461 as the multiple inits of memory domains in that testcase was breaking a lot of stuff.

File: arch/x86/core/x86_mmu.c [hunk 1]
```diff
@@ -646,86 +1054,45 @@ int arch_mem_map(void *virt, uintptr_t phys, size_t size, uint32_t flags)
...
-
-		stack_guard_set(&header->guard_page);
-	} else
-#endif /* CONFIG_USERSPACE */
-	{
-		stack_guard_set(stack);
-	}
+	/* Applied to all page tables as this affects supervisor mode.
+	 * XXX: This never gets reset when the thread exits, which can
+	 * cause problems if the memory is later used for something else.
```
Comment (andrewboie) [inline arch/x86/core/x86_mmu.c:1085]: Bug for this: https://github.com/zephyrproject-rtos/zephyr/issues/29499

File: arch/x86/Kconfig [hunk 2]
```diff
@@ -192,12 +192,38 @@ config X86_MMU
...
+	  Define the number of pages in the pool used to allocate page table
+	  data structures at runtime.
+
+	  Pages might need to be drawn from the pool during memory mapping
+	  operations, unless the address space has been completely pre-allocated.
+
+	  Pages will need to drawn from the pool to initialize memory domains.
+	  This does not include the default memory domain if KPTI=n.
+
+	  The specific value used here depends on the size of physical RAM,
```
Comment (andrewboie) [inline arch/x86/Kconfig:207]: Tuning this is not intuitive. We can do better. I opened https://github.com/zephyrproject-rtos/zephyr/issues/29526

File: arch/x86/core/x86_mmu.c [hunk 3]
```diff
@@ -784,393 +1151,472 @@ int arch_buffer_validate(void *addr, size_t size, int write)
...
 
-	z_x86_apply_mem_domain(thread, thread->mem_domain_info.mem_domain);
+	thread->arch.ptables = (uintptr_t)domain->arch.ptables;
+	LOG_DBG("set thread %p page tables to %p", thread,
+		(void *)thread->arch.ptables);
+
+	/* Check if we're doing a migration from a different memory domain
+	 * and have to remove permissions from its old domain.
+	 *
+	 * XXX: The checks we have to do here and in
```
Comment (andrewboie) [inline arch/x86/core/x86_mmu.c:1569]: Enhancement link for these `is_migration` hacks: https://github.com/zephyrproject-rtos/zephyr/issues/29601

File: arch/x86/core/x86_mmu.c [hunk 4]
```diff
@@ -529,39 +695,117 @@ static void *page_pool_get(void *context)
...
+ * @param ptables Page tables to modify
+ * @param virt Virtual page table entry to update
+ * @param entry_val Value to update in the PTE (ignored if OPTION_RESET)
+ * @param mask What bits to update in the PTE (ignored if OPTION_RESET)
+ * @param options Control options, described above
+ * @retval 0 Success
+ * @retval -ENOMEM allocation required and no free pages (only if OPTION_ALLOC)
  */
 static int page_map_set(pentry_t *ptables, void *virt, pentry_t entry_val,
-			page_get_func_t get_page, void *ctx)
```
Comment (carlocaione) [inline arch/x86/core/x86_mmu.c:562]: You can now remove the typedef on `page_get_func_t`

File: arch/x86/core/x86_mmu.c [hunk 5]
```diff
@@ -784,393 +1151,472 @@ int arch_buffer_validate(void *addr, size_t size, int write)
...
-		return;
+	/* New memory domain we are being added to */
+	struct k_mem_domain *domain = thread->mem_domain_info.mem_domain;
+	/* This is only set for threads that were migrating from some other
+	 * memory domain; new threads this is NULL
+	 */
+	pentry_t *old_ptables = (pentry_t *)thread->arch.ptables;
+	bool is_user = (thread->base.user_options & K_USER) != 0;
+	bool is_migration = (old_ptables != NULL) && is_user;
+			     ;
```
Comment (carlocaione) [inline arch/x86/core/x86_mmu.c:1553]: spurious `;`

Comment (andrewboie) [reply arch/x86/core/x86_mmu.c]: removed.

File: arch/x86/core/x86_mmu.c (hunk 4 above)
Comment (andrewboie) [reply arch/x86/core/x86_mmu.c:562]: should be gone now.

**Review by carlocaione — APPROVED** [review]

File: arch/x86/Kconfig [hunk 6]
```diff
@@ -192,12 +192,26 @@ config X86_MMU
...
+	  much much additional virtual memory will be mapped at runtime, and
```
Comment (dcpleung) [inline arch/x86/Kconfig:208]: "how much"?

**Review by dcpleung — APPROVED** [review]

**Comment by chen-png** [comment]
tested it on acrn x86 platform, it was fine.

**Review by nashif — APPROVED** [review]

**Pull request merged**